Feature/databricks compatibility

fivetran-catfritz opened the pull request:
Pull Request
**Are you a current Fivetran customer?** 
<!--- Please tell us your name, title and company -->
Fivetran created PR

**What change(s) does this PR introduce?** 
<!--- Describe what changes your PR introduces to the package and how to leverage this new feature. -->
PR https://github.com/fivetran/dbt_marketo_source/pull/28 incorporates the following updates:

🚨 Breaking Changes 🚨
Some of the more complex transformation logic has been moved from the Marketo source package to the transform package. This was done so the delineation between staging and intermediate models is in line with Fivetran's other packages. This does not affect the final tables created by the transform package, but this will affect the staging tables as outlined below.

Model stg_marketo__lead_base from dbt_marketo_source has been rolled into [stg_marketo__lead](https://github.com/fivetran/dbt_marketo_source/blob/main/models/stg_marketo__lead.sql).
Parts from model stg_marketo__lead from dbt_marketo_source have been moved to a new model [int_marketo__lead](https://github.com/fivetran/dbt_marketo/blob/feature/create-intermediates/models/intermediate/int_marketo__lead.sql) in dbt_marketo.
Features
🎉 Databricks compatibility 🎉
Ability to disable activity_delete_lead model if necessary (see [README](https://github.com/fivetran/dbt_marketo_source/blob/feature/databricks-compatibility/link) for instructions).
Updated structure of config default variables for enabling campaigns and program models to avoid conflicting with a user's settings.

**Did you update the CHANGELOG?** 
<!--- Please update the new package version’s CHANGELOG entry detailing the changes included in this PR. -->
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [x] Yes

**Does this PR introduce a breaking change?**
<!--- Does this PR introduce changes that will cause current package users' jobs to fail or require a `--full-refresh`? -->
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [x] Yes (please provide breaking change details below.)
- [ ] No  (please provide an explanation as to how the change is non-breaking below.)

**Did you update the dbt_project.yml files with the version upgrade (please leverage standard semantic versioning)? (In both your main project and integration_tests)** 
<!--- The dbt_project.yml and the integration_tests/dbt_project.yml files contain the version number. Be sure to upgrade it accordingly -->
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [x] Yes

**Is this PR in response to a previously created Bug or Feature Request**
<!--- If an Issue was created it is helpful to track the progress by linking it in the PR. -->
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [x] Yes, Issue/Feature #18 #27 
- [ ] No 

**How did you test the PR changes?** 
<!--- Proof of testing is required in order for the PR to be approved. -->
<!--- To check a box, remove the space and insert an x in the box (eg. [x] CircleCi). --> 
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [x] CircleCi <!--- CircleCi testing is only applicable to Fivetran employees. --> 
- [x] Local (please provide additional testing details below)

Tested locally using the below in my `dbt_project.yml` to make sure turning these models on and off worked appropriately after the changes:
``` yml
vars:
    marketo__enable_campaigns:   true    # Default is false
    marketo__enable_programs:     true    # Default is false
    marketo__activity_delete_lead_enabled:  false    # Default is true
```
**Select which warehouse(s) were used to test the PR**
<!--- To check a warehouse remove the space and insert an x in the box (eg. [x] Bigquery). --> 
<!--- To select a checkbox you simply need to add an "x" with no spaces between the brackets (eg. [x] Yes). -->
- [x] BigQuery
- [x] Redshift
- [x] Snowflake
- [x] Postgres
- [x] Databricks
- [ ] Other (provide details below)

**Provide an emoji that best describes your current mood**
<!--- For a complete list of markdown compatible emojis check our this git repo (https://gist.github.com/rxaviers/7360908)  --> 
😕 

**Feedback**

We are so excited you decided to contribute to the Fivetran community dbt package! We continue to work to improve the packages and would greatly appreciate your [feedback](https://www.surveymonkey.com/r/DQ7K7WW) on our existing dbt packages or what you'd like to see next.

fivetran-joemarkiewicz commented on CHANGELOG.md:
```diff
@@ -1,3 +1,14 @@
+# dbt_marketo_source v0.8.0
```
Reminder to add whatever relevant updates from this PR review into the CHANGELOG.

fivetran-catfritz replied:
Done.

fivetran-joemarkiewicz reviewed (changes requested):
@fivetran-catfritz thanks so much for your hard work on these marketo updates! You updates are looking great. I just have a few comments and suggestions that should be addressed before your PR is approved and merged:

- [ ] Address all in-line comments below
- [ ] It seems the GitHub pages are not hosted on this repo. Can you deploy the GitHub pages for our docs and provide the link in the About section. We should have details on how to do this [here](https://fivetran.slab.com/posts/creating-fivetran-dbt-packages-2dc4tt1b#h842b-dbt-docs-site-hosting)
- [x] It looks like the `marketo__activity_delete_lead_enabled` variable is added to the enablement config block within `stg_marketo__activity_delete_lead`, but not in `stg_marketo__activity_delete_lead_tmp`. Can you add the variable to the `_tmp` model as well.
- [ ] Since this will be a breaking change, this may be the best time to add the identifier variables to the `src_marketo.yml` file. Once you add these, be sure to add the respective documentation to the README, and also to update the `integration_tests/dbt_project.yml` to use these variables.
    - This should also result in a CHANGELOG entry.
- [ ] Our updated standards have the source default custom schema to be `[package-name]_source` as opposed to `stg_[package-name]`. Can you update the custom schema config in this package to instead be `marketo_source`.
    - With this update, you will want to also update the README (in both the source and transform) to reference this new default schema. Also, you will want to add this as an entry to the CHANGELOG.

fivetran-joemarkiewicz commented on models/stg_marketo__activity_merge_leads.sql:
```diff
@@ -22,9 +22,9 @@ with base as (
         campaign_id,
-        lead_id,
+        cast(lead_id as {{ dbt_utils.type_int() }}) as lead_id,
         master_updated,
-        trim(trim(merge_ids,']'),'[') as merged_lead_id,
+        cast(replace(trim(trim(merge_ids,']'),'['), ',', ', ') as {{ dbt_utils.type_string() }}) as merged_lead_id,
```
Nice job combining unnecessary downstream steps into one within the staging model!

fivetran-catfritz replied:
Good callout thanks! I think it was a leftover of combining logics, but it makes sense they should both be ints after all the trims. I tested it out and have updated the type of line 27 to int.

fivetran-catfritz replied:
Revision to previous comment--actually it should stay as it was. For line 25 it is a single number, but for line 27 it can be a string of several IDs separated by a comma.

fivetran-catfritz commented on CHANGELOG.md:
```diff
@@ -1,3 +1,14 @@
+# dbt_marketo_source v0.8.0
+PR [28](https://github.com/fivetran/dbt_marketo_source/pull/28) incorporates the following updates:
```
done

fivetran-catfritz commented on dbt_project.yml:
```diff
@@ -25,6 +22,8 @@ vars:
     activity_delete_lead: "{{ source('marketo','activity_delete_lead') }}"
     activity_merge_leads: "{{ source('marketo','activity_merge_leads') }}"
 
+  dbt_utils_dispatch_list: ['spark_utils']
```
I blindly copied it over from another package, but it seems to work without it so I've taken it out.

fivetran-catfritz commented on README.md:
```diff
@@ -1,94 +1,115 @@
+# 📣 What does this dbt package do?
 
-This package enriches your Fivetran data by doing the following:
+- Produces modeled tables that leverage Marketo data from [Fivetran's connector](https://fivetran.com/docs/applications/marketo) in the format described by [this ERD](https://fivetran.com/docs/applications/marketo#schema).
```
Thanks--updated.

fivetran-catfritz commented on README.md:
```diff
@@ -1,94 +1,115 @@
 
-This package enriches your Fivetran data by doing the following:
+- Produces modeled tables that leverage Marketo data from [Fivetran's connector](https://fivetran.com/docs/applications/marketo) in the format described by [this ERD](https://fivetran.com/docs/applications/marketo#schema).
+- Adds descriptions to tables and columns that are synced using Fivetran
+- Models staging tables, which will be used in our transform package
+- Adds column-level testing where applicable. For example, all primary keys are tested for uniqueness and non-null values.
```
Updated.

fivetran-catfritz commented:
@fivetran-joemarkiewicz  In addition to your suggestions I also:
- Added postgres test to integration testing
- Added and revised several column descriptions

With that, this is ready for your review again thanks!

fivetran-catfritz commented on README.md:
```diff
@@ -1,94 +1,115 @@
+- Models staging tables, which will be used in our transform package
+- Adds column-level testing where applicable. For example, all primary keys are tested for uniqueness and non-null values.
+- Generates a comprehensive data dictionary of your source and modeled Marketo data through the [dbt docs site](https://fivetran.github.io/dbt_marketo_source/).
+- These tables are designed to work simultaneously with our [Marketo transformation package](https://https://github.com/fivetran/dbt_marketo_source/).
```
Done.

fivetran-joemarkiewicz reviewed (changes requested):
@fivetran-catfritz thanks so much for making the updates from my initial review! I have just a few minor follow up questions and suggestions. Once you apply those, we should be good to go!

fivetran-joemarkiewicz commented on README.md:
```diff
@@ -1,94 +1,122 @@
+  - Models staging tables, which will be used in our transform package
+  - Adds column-level testing where applicable. For example, all primary keys are tested for uniqueness and non-null values.
+- Generates a comprehensive data dictionary of your source and modeled Marketo data through the [dbt docs site](https://fivetran.github.io/dbt_marketo_source/).
+- These tables are designed to work simultaneously with our [Marketo transformation package](https://https://github.com/fivetran/dbt_marketo/).
```
It looks like you have two `https` instances in this link.

fivetran-catfritz replied:
Thanks done!

fivetran-catfritz commented on README.md:
```diff
@@ -1,94 +1,115 @@
+<br>
+
 ### Changing the Build Schema
 By default this package will build the Marketo staging models within a schema titled (<target_schema> + `_stg_marketo`) in your target database. If this is not where you would like your Marketo data to be written to, add the following configuration to your `dbt_project.yml` file:
```
Sorry about that! Done (actually now).

fivetran-catfritz commented on models/src_marketo.yml:
```diff
@@ -3,10 +3,11 @@ version: 2
   - name: marketo
     schema: "{{ var('marketo_schema', 'marketo') }}"
-    database: "{{ var('marketo_database', target.database) }}"
+    database: "{% if target.type != 'spark'%}{{ var('marketo_database', target.database) }}{% endif %}"
```
Hmm... I realize I don't know the history of the issues we've had with spark, but for Databricks, my understanding is that it can have up to a three-level namespace (metastore/catalog/schema), though for our configuration we only use two (catalog/schema). The databricks "catalog" maps to "database" in dbt from what I can tell. 

The default "database" is `hive_metastore`, but it could possibly have a different name depending on settings. Also I tested a few difference scenarios of providing database and schema variables, and I found even if no value is given for the database, it does still figures out to use the default. However, I'm thinking we should keep the option to set a database value and keep line 6 as it was since it works? Curious your thoughts! It might also be easier for me to show you. Just let me know!

fivetran-catfritz replied:
As we chatted about earlier, for incremental models, I have updated `target.type != 'spark'` to  `target.type not in ['spark', 'databricks']`. For the source yml, I have left as-is.

fivetran-joemarkiewicz commented on CHANGELOG.md:
```diff
@@ -1,3 +1,16 @@
+- Parts from model `stg_marketo__lead` from `dbt_marketo_source` have been moved to a new model [`int_marketo__lead`](https://github.com/fivetran/dbt_marketo/blob/feature/create-intermediates/models/intermediate/int_marketo__lead.sql) in `dbt_marketo`.
+- The default schema for the source tables are now built within a schema titled (`<target_schema>` + `_marketo_source`) in your destination. The previous default schema was (`<target_schema>` + `_stg_marketo`). This may be overwritten if desired.
+## Features
+- 🎉 Databricks compatibility 🎉
```
Since you technically added Postgres compatibility in this update, you should probably highlight that here as well!

fivetran-catfritz replied:
Done!

fivetran-joemarkiewicz reviewed (approved):
@fivetran-catfritz great work on this large scale compatibility update!! I have two super small comments below. Feel free to update those and then we can go forward with the release process!

fivetran-catfritz commented on README.md:
```diff
@@ -1,94 +1,122 @@
 
-This package contains staging models, designed to work simultaneously with our [Marketo transformation package](https://github.com/fivetran/dbt_marketo). The staging models:
+- At least one Fivetran Marketo connector syncing data into your destination.
+- A **BigQuery**, **Snowflake**, **Redshift**, **PostgreSQL**, or **Databricks** destination.
```
Done.
